packager/media/formats/mp2t/es_parser_teletext.cc

```diff
@@ -0,0 +1,512 @@
+  RCHECK(reader.SkipBits(8));
+  std::vector<std::string> lines;
+
+  while (reader.bits_available()) {
+    uint8_t data_unit_id;
+    RCHECK(reader.ReadBits(8, &data_unit_id));
+
+    uint8_t data_unit_length;
+    RCHECK(reader.ReadBits(8, &data_unit_length));
+
+    if (data_unit_length != 44) {
+      LOG(ERROR) << "Bad Teletext data length";
+      break;
+    }
+
+    if (data_unit_id != EBU_TELETEXT_WITH_SUBTITLING) {
+      RCHECK(reader.SkipBytes(44));
+      continue;
+    }
+
+    RCHECK(reader.SkipBits(16));
+
+    uint16_t address_bits;
+    RCHECK(reader.ReadBits(16, &address_bits));
+
+    uint8_t magazine = bit(address_bits, 14) + 2 * bit(address_bits, 12) +
+                       4 * bit(address_bits, 10);
+
+    if (magazine == 0) {
+      magazine = 8;
+    }
+
+    const uint8_t packet_nr =
+        (bit(address_bits, 8) + 2 * bit(address_bits, 6) +
+         4 * bit(address_bits, 4) + 8 * bit(address_bits, 2) +
+         16 * bit(address_bits, 0));
+    const uint8_t* data_block = reader.current_byte_ptr();
+    RCHECK(reader.SkipBytes(40));
+
+    std::string display_text;
+    if (ParseDataBlock(pts, data_block, packet_nr, magazine, display_text)) {
+      lines.emplace_back(std::move(display_text));
+    }
+  }
+
+  if (lines.empty()) {
+    return true;
+  }
+
+  const uint16_t index = magazine_ * 100 + page_number_;
+  auto page_state_itr = page_state_.find(index);
+  if (page_state_itr == page_state_.end()) {
+    page_state_.emplace(index, TextBlock{std::move(lines), {}, last_pts_});
+
+  } else {
+    for (auto& line : lines) {
+      auto& page_state_lines = page_state_itr->second.lines;
+      page_state_lines.emplace_back(std::move(line));
+    }
+    lines.clear();
+  }
+
+  return true;
+}
+
+bool EsParserTeletext::ParseDataBlock(const int64_t pts,
+                                      const uint8_t* data_block,
+                                      const uint8_t packet_nr,
+                                      const uint8_t magazine,
+                                      std::string& display_text) {
+  if (packet_nr == 0) {
+    last_pts_ = pts;
+    BitReader reader(data_block, 32);
+
+    const uint8_t page_number_units = ReadHamming(reader);
+    const uint8_t page_number_tens = ReadHamming(reader);
+    const uint8_t page_number = 10 * page_number_tens + page_number_units;
+
+    const uint16_t index = magazine * 100 + page_number;
+    SendPending(index, pts);
+
+    page_number_ = page_number;
+    magazine_ = magazine;
+    if (page_number == 0xFF) {
+      return false;
+    }
+
+    RCHECK(reader.SkipBits(40));
+    const uint8_t subcode_c11_c14 = ReadHamming(reader);
+    const uint8_t charset_code = subcode_c11_c14 >> 1;
+    if (charset_code != charset_code_) {
+      charset_code_ = charset_code;
+      UpdateCharset();
+    }
+
+    return false;
+
+  } else if (packet_nr == 26) {
+    ParsePacket26(data_block);
+    return false;
+
+  } else if (packet_nr > 26) {
+    return false;
```

Comment:
packet28 has more language codec infos, so it's advisible to parse it and change teletext G0 charset accordingly